cmd/fix.go

```diff
@@ -263,18 +277,122 @@ func fix(args []string, params *fixCommandParams) error {
 		ignore = params.ignoreFiles.v
 	}
 
-	fileProvider := fileprovider.NewFSFileProvider(ignore, args...)
+	filtered, err := config.FilterIgnoredPaths(args, ignore, true, "")
+	if err != nil {
+		return fmt.Errorf("failed to filter ignored paths: %w", err)
+	}
+
+	slices.Sort(filtered)
+	// TODO: Figure out why filtered returns duplicates in the first place
+	filtered = slices.Compact(filtered)
+
+	fileProvider, err := fileprovider.NewInMemoryFileProviderFromFS(filtered...)
+	if err != nil {
+		return fmt.Errorf("failed to create file provider: %w", err)
+	}
 
 	fixReport, err := f.Fix(ctx, &l, fileProvider)
 	if err != nil {
 		return fmt.Errorf("failed to fix: %w", err)
 	}
 
+	gitRepo, err := git.FindGitRepo(args...)
+	if err != nil {
+		return fmt.Errorf("failed to establish git repo: %w", err)
+	}
+
+	// if fix is being run in a git repo, we must not fix files that have been
+	// changed
+	changedFiles := make(map[string]struct{})
+
+	if gitRepo != "" {
+		cf, err := git.GetChangedFiles(gitRepo)
+		if err != nil {
+			return fmt.Errorf("failed to get changed files: %w", err)
+		}
+
+		for _, f := range cf {
+			changedFiles[f] = struct{}{}
+		}
+	}
+
+	var conflictingFiles []string
+
+	for _, file := range fileProvider.ModifiedFiles() {
+		if _, ok := changedFiles[file]; ok {
+			conflictingFiles = append(conflictingFiles, file)
+		}
+	}
+
+	if len(conflictingFiles) > 0 {
+		return fmt.Errorf(
+			`the following files have been changed since the fixer was run:
+- %s
+please run fix from a clean state to support the use of git checkout for undo`,
+			strings.Join(conflictingFiles, "\n- "),
+		)
+	}
+
+	if params.verbose {
+		fmt.Fprintln(outputWriter, "Dry run mode enabled, the following changes would be made:")
+
+		for _, file := range fileProvider.ModifiedFiles() {
+			fmt.Fprintln(outputWriter, "Set:", file, "to:")
+
+			fc, err := fileProvider.GetFile(file)
+			if err != nil {
+				return fmt.Errorf("failed to get file %s: %w", file, err)
+			}
+
+			fmt.Fprintln(outputWriter, string(fc))
+			fmt.Fprintln(outputWriter, "----------")
+		}
+
+		for _, file := range fileProvider.DeletedFiles() {
+			fmt.Fprintln(outputWriter, "Delete:", file)
+		}
+	}
+
+	if !params.dryRun {
+		for _, file := range fileProvider.DeletedFiles() {
+			err := os.Remove(file)
+			if err != nil {
+				return fmt.Errorf("failed to delete file %s: %w", file, err)
+			}
+		}
+
+		for _, file := range fileProvider.ModifiedFiles() {
```

Comment:
Something like a `fileprovider.WriteToDisk()` might be a nice abstraction.

Reply:
Yeah, this is true. I guess that I'd count this as in the 'to refactor' list for now, we only have one file provider, and it's in memory. It wasn't immediately clear how this func would work relative paths etc.